primitives/arithmetic/src/fixed.rs

```diff
@@ -0,0 +1,1311 @@
+	/// 
+	/// Returns `Self::max` or `Self::min` if `int` exceeds accuracy.
+	fn from_integer<N: UniqueSaturatedInto<Self::Inner>>(int: N) -> Self;
+
+	/// Creates `self` from an integer number `int`.
+	/// 
+	/// Returns `None` if `int` exceeds accuracy.
+	fn checked_from_integer(int: Self::Inner) -> Option<Self>;
+
+	/// Creates `self` from a rational number. Equal to `n / d`.
+	///
+	/// Panics if `d = 0`. Returns `Self::max` or `Self::min` if `n / d` exceeds accuracy.
+	fn from_rational<N: FixedPointOperand, D: FixedPointOperand>(n: N, d: D) -> Self;
+
+	/// Creates `self` from a rational number. Equal to `n / d`.
+	///
+	/// Returns `None` if `d == 0` or `n / d` exceeds accuracy.
+	fn checked_from_rational<N: FixedPointOperand, D: FixedPointOperand>(n: N, d: D) -> Option<Self>;
+
+	/// Checked multiplication for integer type `N`. Equal to `self * n`.
+	///
+	/// Returns `None` if the result does not fit in `N`.
+	fn checked_mul_int<N: FixedPointOperand>(self, n: N) -> Option<N>;
+
+	/// Saturating multiplication for integer type `N`. Equal to `self * n`.
+	///
+	/// Returns `N::min` or `N::max` if the result does not fit in `N`.
+	fn saturating_mul_int<N: FixedPointOperand>(self, n: N) -> N;
+
+	/// Checked division for integer type `N`. Equal to `self / d`.
+	///
+	/// Returns `None` if the result does not fit in `N` or `d == 0`.
+	fn checked_div_int<N: FixedPointOperand>(self, d: N) -> Option<N>;
+
+	/// Saturating division for integer type `N`. Equal to `self / d`.
+	///
+	/// Panics if `d == 0`. Returns `N::min` or `N::max` if the result does not fit in `N`.
+	fn saturating_div_int<N: FixedPointOperand>(self, d: N) -> N;
+
+	/// Saturating multiplication for integer type `N`, adding the result back.
+	/// Equal to `self * n + n`.
+	///
+	/// Returns `N::min` or `N::max` if the multiplication or final result do not fit in `N`.
+	fn saturating_mul_int_acc<N: FixedPointOperand>(self, n: N) -> N;
+
+	/// Saturating absolute value.
+	///
+	/// Returns `Self::max` if `self == Self::min`.
+	fn saturating_abs(self) -> Self;
+
+	/// Takes the reciprocal (inverse). Equal to `1 / self`.
+	///
+	/// Returns `None` if `self = 0`.
+	fn reciprocal(self) -> Option<Self> {
+		Self::one().checked_div(&self)
+	}
+
+	/// Returns zero.
+	fn zero() -> Self;
+
+	/// Checks if the number is zero.
+	fn is_zero(&self) -> bool;
+
+	/// Returns one.
+	fn one() -> Self;
+
+	/// Checks if the number is positive.
+	fn is_positive(self) -> bool;
+
+	/// Checks if the number is negative.
+	fn is_negative(self) -> bool;
+
+	/// Returns the integer part.
+	fn trunc(self) -> Self;
+
+	/// Returns the fractional part.
+	fn frac(self) -> Self;
+
+	/// Returns the smallest integer greater than or equal to a number.
+	fn ceil(self) -> Self;
+
+	/// Returns the largest integer less than or equal to a number.
+	fn floor(self) -> Self;
+
+	/// Returns the nearest integer to a number. Round half-way cases away from 0.0.
+	fn round(self) -> Self;
```

Comment:
so are these always accurate? even near all the theoretical limits?

Reply:
Some of them saturate, I should add the `saturating_` probably.